Add Cypress Rules documentation (#6215)
cypress-io/cypress-documentation: docs/accessibility/core-concepts/cypress-rules.mdx

**Rule ID**: `cy-semantic-html-warning`

<DocsImage
  src="/img/accessibility/core-concepts/cypress-rule.png"
  width="60%"
  alt="A custom Cypress accessibility rule called 'Interactive elements must be semantically correct' has two failed elements displayed. One element is open and message reads 'This element should use semantic HTML or ARIA to support the interactions that take place during tests. Detected interactions: mousedown, mouseup, click."
/>

This custom rule is intended to surface potential usability problems with the kinds of elements that Axe-Core® traditionally does not evaluate as interactive, such as div, span, SVG, and image elements.

This works based on the interactions that take place during testing. Often an inaccessible custom component will have Cypress tests that perform interactions the way a user does, and that allows us to detect a mismatch between the underlying purpose of the element in code, and the usage of the element in the context of a test.

### How to pass this rule

You should look at the purpose and nature of the specific elements flagged by this rule and decide what interactive element is appropriate.

For example:

> - If the element that gets flagged is a span that has been styled to look like a button and submits a form, the correct solution would be standard HTML button element. This communicates the expected action and is activated with the space bar or enter key.

Reviewer: nit: something missing here

```suggestion
- If the element that gets flagged is a span that has been styled to look like a button and submits a form, the correct solution would be to instead use a standard HTML button element. This communicates the expected action and is activated with the space bar or enter key.
```

> - If the same element is styled with underlined text, and when clicked it resets the URL to navigate to a new page, the correct solution is a link element with a `href` attribute.

Reviewer: nit: dissimilar phrasing on the point above

```suggestion
- If the same element is styled with underlined text, and when clicked it resets the URL to navigate to a new page, the correct solution would be a link element with a `href` attribute.
```

---
title: 'Cypress Rules | Cypress Accessibility Documentation'
description: 'Review the main areas to pay attention to when first reviewing an accessibility report for a Cypress run.'
sidebar_position: 70
sidebar_label: Cypress rules
sidebar_custom_props: { 'new_label': true }
---

<ProductHeading product="accessibility" />

# Cypress rules

> In addition to running the default ruleset of the Axe-Core® library, Cypress develops custom rules that take advantage of the addition layer of information available in a test automation context. These rules are identified by a "Cypress Rule" badge, to distinguish them from the main Axe-Core® rule set.

Reviewer 2: ```suggestion
In addition to running the default ruleset of the Axe-Core® library, Cypress offers custom rules that take advantage of the additional layer of information available in the test automation context. These rules are identified by a "Cypress Rule" badge, to distinguish them from the main Axe-Core® rule set.
```

For example:

- If the element that gets flagged is a span that has been styled to look like a button and submits a form, the correct solution would be to instead use a standard HTML button element. This communicates the expected action and is activated with the space bar or enter key.

- If the same element is styled with underlined text, and when clicked it resets the URL to navigate to a new page, the correct solution would be a link element with a `href` attribute.

In each case, the solution is to choose the element that provides the best information to the user, so they can independently understand what the action does, and activate it.

After you have updated to a suitable HTML solution, elements will then be able to be correctly parsed and analyzed by the standard Axe-Core® checks.

:::info
**Tip**: Follow the [first rule of ARIA use](https://www.w3.org/TR/using-aria/#firstrule) and implement interactive controls with standard semantic HTML where possible, relying on ARIA `role` attributes only when no suitable HTML element is available.
:::

### Why this matters

> In the Web Content Accessibility Guidelines (WCAG), interactive elements are required to met a wide range of Success Criteria (SCs) in order to follow the POUR principles, which stands for Perceivable, Operable, Understandable, and Robust.

Reviewer 2: ```suggestion
In the Web Content Accessibility Guidelines (WCAG), interactive elements are required to meet a wide range of Success Criteria (SCs) in order to follow the POUR principles, which stands for Perceivable, Operable, Understandable, and Robust.
```

Interactive elements that are not semantically correct often don't appear in automated scans, but do appear in manual testing, because they are usually not operable with a keyboard or correctly understandable when described by a screen reader.

> Custom components implemented this way have a direct impact on the independence of your disabled users.

Reviewer 2: ```suggestion
Custom components implemented this way have a direct impact on the independence of disabled users.
```
